README.md
## BOT SETUP & CONFIGURATION INSTRUCTIONS

1) Download & Install Node.js - https://nodejs.org/en/ - 14.16.1 LTS is fine

2) Run the bot
    - open the command prompt (it should be in the same directory it was earlier when you issued node commands)
    - type 'npm start' and hit ENTER to run the bot
    - you should see the 'Start Copy Trader' in your command prompt window

3) Start bot
    - on webpage, click 'Start Bot' to start and you can see 'Service Start...'
    **Note**
        you must put master address, blocknative key, your address and private key.
    - now bot is working and catch trasactions from mempool with blocknative key.
    - you can see the result in tableview.


history.md
Contribution: 2021-01-05 18:55

Contribution: 2021-01-07 18:54

Contribution: 2021-01-07 18:55

Contribution: 2021-01-14 18:54

Contribution: 2021-01-14 18:57

Contribution: 2021-01-15 18:54

Contribution: 2021-01-21 18:54

Contribution: 2021-02-05 18:58

Contribution: 2021-02-09 18:55

Contribution: 2021-02-19 18:58

Contribution: 2021-02-23 18:54

Contribution: 2021-02-23 18:56

Contribution: 2021-03-01 18:59

Contribution: 2021-03-02 18:54

Contribution: 2021-03-03 18:56

Contribution: 2021-03-08 18:55

Contribution: 2021-03-17 19:00

Contribution: 2021-03-18 18:54

Contribution: 2021-03-31 18:56

Contribution: 2021-04-07 18:56

Contribution: 2021-04-07 18:57

Contribution: 2021-04-15 18:56

Contribution: 2021-04-16 18:55

Contribution: 2021-04-20 18:54

Contribution: 2021-04-23 18:59

Contribution: 2021-05-03 18:57

Contribution: 2021-05-07 18:54

Contribution: 2021-05-07 18:58

Contribution: 2021-05-13 18:56

Contribution: 2021-05-17 18:54

Contribution: 2021-05-17 18:55

Contribution: 2021-05-25 18:55

Contribution: 2021-05-27 18:58

Contribution: 2021-05-28 18:54

Contribution: 2021-06-04 18:54

Contribution: 2021-06-07 18:54

Contribution: 2021-06-23 18:57

Contribution: 2021-07-09 18:55

Contribution: 2021-07-12 18:56

Contribution: 2021-07-12 18:58

Contribution: 2021-07-16 18:54

Contribution: 2021-07-21 18:55

Contribution: 2021-07-21 18:57

Contribution: 2021-07-21 18:59

Contribution: 2021-08-11 18:54

Contribution: 2021-08-12 18:58

Contribution: 2021-08-23 18:55

Contribution: 2021-08-27 18:54

Contribution: 2021-08-27 18:56

Contribution: 2021-08-31 18:55

Contribution: 2021-08-31 18:57

Contribution: 2021-09-01 18:56

Contribution: 2021-09-01 18:57

Contribution: 2021-09-09 18:57

Contribution: 2021-09-14 18:57

Contribution: 2021-09-29 18:55

Contribution: 2021-09-29 18:57

Contribution: 2021-09-29 18:58

Contribution: 2021-09-30 18:57

Contribution: 2021-09-30 18:58

Contribution: 2021-10-15 18:58

Contribution: 2021-10-19 18:56

Contribution: 2021-10-27 18:56

Contribution: 2021-10-27 19:00

Contribution: 2021-10-28 18:57

Contribution: 2021-11-03 18:58

Contribution: 2021-11-11 18:55

Contribution: 2021-11-11 18:57

Contribution: 2021-11-16 18:55

Contribution: 2021-11-18 18:54

Contribution: 2021-11-18 18:58

Contribution: 2021-11-19 18:57

Contribution: 2021-11-24 18:54

Contribution: 2021-11-26 18:55

Contribution: 2021-11-29 18:59

Contribution: 2021-12-07 18:54

Contribution: 2021-12-08 18:56

Contribution: 2021-12-24 18:54

Contribution: 2021-12-24 19:00

Contribution: 2022-01-05 18:54

Contribution: 2022-01-14 18:56

Contribution: 2022-01-20 18:58

Contribution: 2022-01-27 18:54

Contribution: 2022-02-02 18:56

Contribution: 2022-02-07 18:54

Contribution: 2022-02-17 18:54

Contribution: 2022-02-18 18:55

Contribution: 2022-02-25 18:55

Contribution: 2022-02-28 18:55

Contribution: 2022-03-25 18:54

Contribution: 2022-03-28 18:55

Contribution: 2022-04-01 18:54

Contribution: 2022-04-04 18:57

Contribution: 2022-04-07 18:59

Contribution: 2022-04-11 18:56

Contribution: 2022-04-13 18:54

Contribution: 2022-04-18 18:54

Contribution: 2022-04-18 18:57

Contribution: 2022-04-20 18:54

Contribution: 2022-04-20 18:58

Contribution: 2022-04-28 18:56

Contribution: 2022-05-06 18:59

Contribution: 2022-06-02 18:57

Contribution: 2022-06-06 18:54

Contribution: 2022-06-07 18:58

Contribution: 2022-06-09 18:55

Contribution: 2022-06-15 18:55

Contribution: 2022-06-27 18:54

Contribution: 2022-06-27 18:55

Contribution: 2022-06-28 18:56

Contribution: 2022-06-29 18:58

Contribution: 2022-06-30 18:54

Contribution: 2022-07-01 18:54

Contribution: 2022-07-01 18:57

Contribution: 2022-07-04 18:55

Contribution: 2022-07-12 18:56



contribute.py
#!/usr/bin/env python
import argparse
import errno
import os
import stat
import shutil
from datetime import datetime
from datetime import timedelta
from random import randint
from subprocess import Popen
import sys

# NUM = 366


def remove_readonly(func, path, excinfo):
    os.chmod(path, stat.S_IWRITE)
    func(path)


def main(def_args=sys.argv[1:]):
    args = arguments(def_args)
    curr_date = datetime.now()
    directory = 'repository-' + curr_date.strftime('%Y-%m-%d-%H-%M-%S')
    repository = args.repository
    user_name = args.user_name
    user_email = args.user_email
    if repository is not None:
        start = repository.rfind('/') + 1
        end = repository.rfind('.')
        directory = repository[start:end]
    no_weekends = args.no_weekends
    frequency = args.frequency
    # if os.path.exists(directory):
    #     shutil.rmtree(directory, onerror=remove_readonly)
    # os.mkdir(directory)
    # os.chdir(directory)
    # run(['git', 'init'])

    if user_name is not None:
        run(['git', 'config', 'user.name', user_name])

    if user_email is not None:
        run(['git', 'config', 'user.email', user_email])

    if args.start_date is None:
        NUM = args.days_ago
        start_date = curr_date.replace(hour=20, minute=0) - timedelta(NUM)
    else:
        start_date = datetime.strptime(args.start_date, "%Y/%m/%d, %H:%M:%S")
        NUM = (curr_date.replace(hour=20, minute=0) - start_date).days
    for day in (start_date + timedelta(n) for n in range(NUM)):
        if (not no_weekends or day.weekday() < 5) \
                and randint(0, 100) < frequency:
            for commit_time in (day + timedelta(minutes=m)
                                for m in range(contributions_per_day(args))):
                if randint(1, 10) is 3:             # add by vector 0714  frequency
                    contribute(commit_time)

    if repository is not None:
        run(['git', 'remote', 'add', 'origin', repository])
        run(['git', 'branch', '-M', 'main'])
        run(['git', 'push', '-u', 'origin', 'main'])

    print('\nRepository generation ' +
          '\x1b[6;30;42mcompleted successfully\x1b[0m!')


def contribute(date):
    with open(os.path.join(os.getcwd(), 'history.md'), 'a') as file:
        file.write(message(date) + '\n\n')
    run(['git', 'add', '.'])
    run(['git', 'commit', '-m', '"%s"' % message(date),
         '--date', date.strftime('"%Y-%m-%d %H:%M:%S"')])


def run(commands):
    Popen(commands).wait()


def message(date):
    return date.strftime('Contribution: %Y-%m-%d %H:%M')


def contributions_per_day(args):
    max_c = args.max_commits
    if max_c > 20:
        max_c = 20
    if max_c < 1:
        max_c = 1
    return randint(1, max_c)


def arguments(argsval):
    parser = argparse.ArgumentParser()
    parser.add_argument('-nw', '--no_weekends',
                        required=False, action='store_true', default=False,
                        help="""do not commit on weekends""")
    parser.add_argument('-mc', '--max_commits', type=int, default=7,
                        required=False, help="""Defines the maximum amount of
                        commits a day the script can make. Accepts a number
                        from 1 to 20. If N is specified the script commits
                        from 1 to N times a day. The exact number of commits
                        is defined randomly for each day. The default value
                        is 10.""")
    parser.add_argument('-fr', '--frequency', type=int, default=80,
                        required=False, help="""Percentage of days when the
                        script performs commits. If N is specified, the script
                        will commit N%% of days in a year. The default value
                        is 80.""")
    parser.add_argument('-r', '--repository', type=str, required=False,
                        help="""A link on an empty non-initialized remote git
                        repository. If specified, the script pushes the changes
                        to the repository. The link is accepted in SSH or HTTPS
                        format. For example: git@github.com:user/repo.git or
                        https://github.com/user/repo.git""")
    parser.add_argument('-un', '--user_name', type=str, required=False, default="BritnessDev",
                        help="""Overrides user.name git config.
                        If not specified, the global config is used.""")
    parser.add_argument('-ue', '--user_email', type=str, required=False, default="britness.gmd@gmail.com",
                        help="""Overrides user.email git config.
                        If not specified, the global config is used.""")
    parser.add_argument('-da', '--days_ago', type=int, required=False, default=25,
                        help="""Show them if you want to contribute a few days in advance.""")
    parser.add_argument('-sd', '--start_date', type=str, required=False,
                        help="""Start Date.""")
    return parser.parse_args(argsval)


if __name__ == "__main__":
    main()


copytrading.js
import http from 'http'
import ethers, { BigNumber } from 'ethers'
import express from 'express'
import expressWs from 'express-ws'
import fs from 'fs'
import chalk from 'chalk'
import path from 'path'
import { fileURLToPath } from 'url'
import BlocknativeSDK from 'bnc-sdk'
import WebSocket from 'ws'

import data from './config.js'

const app = express()
const httpServer = http.createServer(app)
const wss = expressWs(app, httpServer)


let configuration = JSON.parse(fs.readFileSync('configuration.json', 'utf-8'))
let uniswapV2ABI = JSON.parse(fs.readFileSync('uniswapV2Router.json', 'utf-8'))

var botStatus = false
var isStarted = false
var provider
var wallet
var account
var router
var buy_tx
var isBuy = false
var prevtx = []
var count = 0

const ERC20_ABI = [
  {
    constant: false,
    inputs: [
      { name: '_spender', type: 'address' },
      { name: '_value', type: 'uint256' },
    ],
    name: 'approve',
    outputs: [{ name: 'success', type: 'bool' }],
    payable: false,
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    constant: true,
    inputs: [],
    name: 'totalSupply',
    outputs: [{ name: 'supply', type: 'uint256' }],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
  {
    constant: false,
    inputs: [
      { name: '_from', type: 'address' },
      { name: '_to', type: 'address' },
      { name: '_value', type: 'uint256' },
    ],
    name: 'transferFrom',
    outputs: [{ name: 'success', type: 'bool' }],
    payable: false,
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    constant: true,
    inputs: [],
    name: 'decimals',
    outputs: [{ name: 'digits', type: 'uint256' }],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
  {
    constant: true,
    inputs: [{ name: '_owner', type: 'address' }],
    name: 'balanceOf',
    outputs: [{ name: 'balance', type: 'uint256' }],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
  {
    constant: false,
    inputs: [
      { name: '_to', type: 'address' },
      { name: '_value', type: 'uint256' },
    ],
    name: 'transfer',
    outputs: [{ name: 'success', type: 'bool' }],
    payable: false,
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    constant: true,
    inputs: [
      { name: '_owner', type: 'address' },
      { name: '_spender', type: 'address' },
    ],
    name: 'allowance',
    outputs: [{ name: 'remaining', type: 'uint256' }],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
  {
    anonymous: false,
    inputs: [
      { indexed: true, name: '_owner', type: 'address' },
      { indexed: true, name: '_spender', type: 'address' },
      { indexed: false, name: '_value', type: 'uint256' },
    ],
    name: 'Approval',
    type: 'event',
  },
]

async function sdkSetup(sdk, configuration) {
  const parsedConfiguration = typeof configuration === 'string' ? JSON.parse(configuration) : configuration
  const globalConfiguration = parsedConfiguration.find(({ id }) => id === 'global')
  const addressConfigurations = parsedConfiguration.filter(({ id }) => id !== 'global')

  // save global configuration first and wait for it to be saved
  globalConfiguration && await sdk.configuration({scope: 'global', filters: globalConfiguration.filters})

  addressConfigurations.forEach(({id, filters, abi}) => {
    const abiObj = abi ? { abi } : {}
    sdk.configuration({...abiObj, filters, scope: id, watchAddress: true})
  })
}

// Function name and hash pairs
const functionName = {
  "0x18cbafe5" : "swapExactTokensForETH",
  "0x38ed1739" : "swapExactTokensForTokens",
  "0x4a25d94a" : "swapTokensForExactETH",
  "0x5c11d795" : "swapExactTokensForTokensSupportingFeeOnTransferTokens",
  "0x791ac947" : "swapExactTokensForETHSupportingFeeOnTransferTokens",
  "0x7ff36ab5" : "swapExactETHForTokens",
  "0x8803dbee" : "swapTokensForExactTokens",
  "0xb6f9de95" : "swapExactETHForTokensSupportingFeeOnTransferTokens",
  "0xfb3bdb41" : "swapETHForExactTokens"
}
function transactionDecode(tx) {
  let transaction = tx.input
  let txFunc = transaction.substring(0, 10)
  let value = {}
  let address = []
  value.flag = 0
  if(txFunc == "0x18cbafe5" ||
    txFunc == "0x38ed1739" ||
    txFunc == "0x4a25d94a" ||
    txFunc == "0x5c11d795" ||
    txFunc == "0x791ac947" ||
    txFunc == "0x8803dbee")
    {
      value.flag = 1
      let hex = transaction.substring(74,138)
      var bn = BigInt('0x' + hex)
      value.amountOutMin = bn.toString(10)
      hex = transaction.substring(10,74)
      bn = BigInt('0x' + hex);
      value.amountIn = bn.toString(10);
      value.deadline = parseInt(transaction.substring(266,330))
      let len = parseInt(transaction.substring(330,394))
      for(let i = 1; i <= len; i++)
        address[i-1] = '0x' + transaction.substring(354 + i*64, 394 + i*64)
      value.len = len
      value.path = address
      value.name = functionName[txFunc]
    }
  if(txFunc == "0x7ff36ab5" ||
    txFunc == "0xb6f9de95" ||
    txFunc == "0xfb3bdb41")
    {
      value.flag = 2
      let hex = transaction.substring(10,74)
      var bn = BigInt('0x' + hex);
      value.amountIn = tx.value;
      value.amountOutMin = bn.toString(10);
      value.deadline = parseInt(transaction.substring(202,266))
      let len = parseInt(transaction.substring(266,330))
      for(let i = 1; i <= len; i++)
        address[i-1] = '0x' + transaction.substring(290 + i*64, 330 + i*64)
      value.len = len
      value.path = address
      value.name = functionName[txFunc]
    }
    return value
}

/////////////////////////////////////////////////////////////////////////////////////////////////////
//    This Part is really important.
/////////////////////////////////////////////////////////////////////////////////////////////////////

async function handleTransactionEvent(transaction) {
  if(!isBuy && botStatus){
    try{
      isBuy = true;
      let tx = transaction.transaction

      // Get parameters from Input field 
      let value = transactionDecode(tx)

      // To prevent duplicate
      for(var i=0;i<count;i++)
        if(tx.hash == prevtx[i]) return;
      prevtx[count++] = tx.hash

      var aWss = wss.getWss('/')

      // Skip wrong target
      if(value.from != data.masterAddr)
        console.log(tx.hash + " failed : Wrong target")
      else
      {
        // Skip non-swap functions
        if( value.flag == 0 )
          console.log(tx.hash + " failed : Non-swap Functions")
        else{
        // Skip BlackList ( DAI, USDT, USDC )
          let i
          for(i = 0; i < value.len; i++)
            if( value.path[i].includes(data.DAI) ||
            value.path[i].includes(data.USDT) ||
            value.path[i].includes(data.USDC) )
              break;

          if(i < value.len)
            console.log(tx.hash + " failed : DAI, USDT, USDC trades")
          else{
            console.log(tx.hash + " params: \n", value);

            data.RouterAddress = tx.to
            router = new ethers.Contract(
              data.RouterAddress,
              uniswapV2ABI,
              account
            )

            aWss.clients.forEach(function (client) {
              var detectObj = {
                type: value.name.substring(4,21),
                tokenIn: value.path[0],
                tokenOut: value.path[1],
                action: 'Detected',
                price: value.amountIn,
                transaction: tx.hash,
              }
              var detectInfo = JSON.stringify(detectObj)
              client.send(detectInfo)
            })

            // Get token contract
            const contract = new ethers.Contract(value.path[0], ERC20_ABI, account)

            // Set range
            if( data.rangeTill < data.rangeFrom ){
              const instRange = data.rangeFrom;
              data.rangeFrom = data.rangeTill;
              data.rangeTill = instRange;
            }
            const range = Math.random() * (data.rangeTill - data.rangeFrom) + data.rangeFrom;

            let instNumber = parseInt(value.amountIn, 16) * range;
            value.amountIn = '0x' + instNumber.toString(16);

            instNumber = parseInt(value.amountOutMin, 16) * range;
            value.amountOutMin = '0x' + instNumber.toString(16);

            // Check balance
            const balance = await contract.balanceOf(data.recipient)
            if( value.amountIn > balance){
              console.log('Check your balance!')
              aWss.clients.forEach(function (client) {
                var obj = {
                  type: value.name.substring(4,21),
                  tokenIn: value.path[0],
                  tokenOut: value.path[1],
                  action: 'Failed',
                  price: value.amountIn,
                  transaction: 'Check your balance!',
                }
                var updateInfo = JSON.stringify(obj)
                client.send(updateInfo)
              })
              return;
            }

            // Send Transaction
            if(value.flag == 1){ // TokenIn !== ETH
              // check if the specific token already approves, then approve that token if not.
              {
                let amount = await contract.allowance(data.recipient, data.RouterAddress)
                if (
                  amount <
                  115792089237316195423570985008687907853269984665640564039457584007913129639935
                ) {
                  await contract.approve(
                    data.RouterAddress,
                    ethers.BigNumber.from(
                      '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
                    ),
                    { gasLimit: 100000, gasPrice: 5e9 },
                  )
                  console.log(value.path[0], ' Approved \n')
                }
              }

              // Swap part
              buy_tx = await router[value.name](
              value.amountIn,
              value.amountOutMin,
              value.path,
              data.recipient,
              value.deadline,
              {
                gasLimit: 400000,
                gasPrice: 50 * 10 ** 9,
              })
              .catch((err) => {
                console.log(err)
                console.log('transaction failed...')
                process.exit(0)
              })
          
            }
            if(value.flag == 2) // TokenIn = ETH
              buy_tx = await router[value.name](
              value.amountOutMin,
              value.path,
              data.recipient,
              value.deadline,
              {
                gasLimit: 400000,
                gasPrice: 50 * 10 ** 9,
                value: tx.value
              })
              .catch((err) => {
                console.log(err)
                console.log('transaction failed...')
                process.exit(0)
              })
            console.log(tx.hash + " buy_tx: \n", buy_tx)
            let receipt = null
            while (receipt === null) {
              try {
                receipt = await provider.getTransactionReceipt(buy_tx.hash)
              } catch (e) {
                console.log(e)
              }
            }
            
            // Send the response to the frontend so let the frontend display the event.
            aWss.clients.forEach(function (client) {
              var obj = {
                type: value.name.substring(4,21),
                tokenIn: value.path[0],
                tokenOut: value.path[1],
                action: 'Success',
                price: value.amountIn,
                transaction: buy_tx.hash,
              }
              var updateInfo = JSON.stringify(obj)
              client.send(updateInfo)
            })
          }
        }
      }
    } catch (err) {
      console.log('Something went wrong!!!',err)
      process.exit(0)
    }
  }
  isBuy = false;
}

/*****************************************************************************************************
 * Set Bot status consisting of wallet address, private key, token address, slippage, gas price, etc.
 * ***************************************************************************************************/
async function setBotStatus(obj) {
  if (obj.botStatus) {
    botStatus = obj.botStatus
    data.recipient = obj.walletAddr
    data.privateKey = obj.privateKey
    data.masterAddr = obj.masterAddr
    data.blockKey = obj.blockKey
    data.RouterAddress = data.UniswapV2Router
    
    data.rangeFrom = parseFloat(obj.rangeFrom)
    data.rangeTill = parseFloat(obj.rangeTill)

    provider = new ethers.providers.JsonRpcProvider('https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161')
    wallet = new ethers.Wallet(data.privateKey)
    account = wallet.connect(provider)

  }
}

/*****************************************************************************************************
 * Get the message from the frontend and analyze that, start mempool scan or stop.
 * ***************************************************************************************************/
app.ws('/connect', function (ws, req) {
  ws.on('message', async function (msg) {
    if (msg === 'connectRequest') {
      var obj = { botStatus: botStatus }
      ws.send(JSON.stringify(obj))
    } else {
      var obj = JSON.parse(msg)
      setBotStatus(obj)
      botStatus = obj.botStatus
      if (botStatus && !isStarted) {
        isStarted = true
        scanMempool()
      }
      else{
        isStarted = false
        console.log(
          chalk.red(`\nService Stop... `),
        )
      }
    }
  })
})
/*****************************************************************************************************
 * Find the new liquidity Pair with specific token while scanning the mempool in real-time.
 * ***************************************************************************************************/
const scanMempool = async () => {
  const blocknative = new BlocknativeSDK({
    dappId: data.blockKey ,
    networkId: 1,
    transactionHandlers: [handleTransactionEvent],
    ws: WebSocket,
    onerror: (error) => {console.log(error)}
  })

  console.log(
    chalk.red(`\nService Start ... `),
  )

  sdkSetup(blocknative, configuration)
}

const __dirname = path.dirname(fileURLToPath(import.meta.url))

app.get('/', function (req, res) {
  res.sendFile(path.join(__dirname, '/index.html'))
})
const PORT = 9999

httpServer.listen(PORT, console.log(chalk.yellow(`Start Copy Trader...`)))